flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java

```diff
@@ -283,15 +284,30 @@ public ArchivedExecution getPriorExecutionAttempt(int attemptNumber) {
         }
     }
 
-    public ArchivedExecution getLatestPriorExecution() {
-        synchronized (priorExecutions) {
-            final int size = priorExecutions.size();
-            if (size > 0) {
-                return priorExecutions.get(size - 1);
-            } else {
-                return null;
+    /**
+     * Gets the latest property from a prior execution that is not null.
+     *
+     * @param extractor defining the property to extract
+     * @param <T> type of the property
+     * @return Optional containing the latest property if it exists; otherwise {@code
+     *     Optional.empty()}.
+     */
+    private <T> Optional<T> getLatestPriorProperty(Function<ArchivedExecution, T> extractor) {
```

Comment:
just wondering, body of the this method used to be synchronized, is it safe to remove that?

Reply:
Yes. The `ExecutionGraph` used to be thread safe but not it is only driven by a single thread. This made many things a lot easier.

Reply:
Should we remove the synchronization from other places as well then? eg. `ExecutionVertex#getPriorExecutionAttempt`

Reply:
Eventually, we should remove it, yes.